iocage/lib/LaunchableResourceUpdate.py

```diff
@@ -0,0 +1,405 @@
+        now = datetime.datetime.utcnow()
+        text += now.strftime("%Y%m%d%H%I%S.%f")
+        return text
+
+    def fetch(
+        self
+    ) -> typing.Generator['iocage.lib.events.IocageEvent', None, None]:
+
+        events = iocage.lib.events
+        releaseUpdatePullEvent = events.ReleaseUpdatePull(self.release)
+        releaseUpdateDownloadEvent = events.ReleaseUpdateDownload(self.release)
+
+        yield releaseUpdatePullEvent.begin()
+        try:
+            self._pull_updater()
+        except Exception as e:
+            yield releaseUpdatePullEvent.fail(e)
+            raise          
+        yield releaseUpdatePullEvent.end()
+
+        yield releaseUpdateDownloadEvent.begin()
+        self.logger.verbose(
+            f"Fetching updates for release '{self.release.name}'"
+        )
+        try:
+            iocage.lib.helpers.exec(
+                self._fetch_command,
+                logger=self.logger
+            )
+        except Exception as e:
+            yield releaseUpdateDownloadEvent.fail(e)
+            raise
+        yield releaseUpdateDownloadEvent.end()
+
+    def apply(
+        self
+    ) -> typing.Generator['iocage.lib.events.IocageEvent', None, None]:
+
+        dataset = self.resource.dataset
+        snapshot_name = self._append_datetime(f"{dataset.name}@pre-update")
+
+        runReleaseUpdateEvent = iocage.lib.events.RunReleaseUpdate(
+            self.release
+        )
+        yield runReleaseUpdateEvent.begin()
+
+        # create snapshot before the changes
+        dataset.snapshot(snapshot_name, recursive=True)
+
+        jail = self.temporary_jail
+        changed = False
+
+        try:
+            for event in self._update_jail(jail):
+                if isinstance(event, iocage.lib.events.IocageEvent):
+                    yield event
+                else:
+                    changed = event
+        except Exception as e:
+            # kill the helper jail and roll back if anything went wrong
+            self.logger.verbose(
+                "There was an error updating the Jail - reverting the changes"
+            )
+            jail.stop(force=True)
+            self.resource.zfs.get_snapshot(snapshot_name).rollback(force=True)
+            yield runReleaseUpdateEvent.fail(e)
+            raise e
+
+        jail.stop()
+        yield runReleaseUpdateEvent.end()
+        yield changed
+
+    def _update_jail(
+        self,
+        jail: 'iocage.lib.Jail.JailGenerator'
+    ) -> typing.Generator['iocage.lib.events.IocageEvent', None, None]:
+
+        events = iocage.lib.events
+        executeReleaseUpdateEvent = events.ExecuteReleaseUpdate(self.release)
+
+        try:
+            for event in jail.fork_exec(self._update_command):
+                yield event
+                self.logger.debug(
+                    f"Update of release '{self.release.name}' finished"
+                )
+        except Exception:
+            raise
```

Comment:
this raise without arguments probably needs to go, in favour of what we already have in the next line